ports/atmel-samd/tick.c

```diff
@@ -131,3 +132,29 @@ void wait_until(uint64_t ms, uint32_t us_until_ms) {
     uint32_t ticks_per_us = common_hal_mcu_processor_get_frequency() / 1000 / 1000;
     while (ticks_ms <= ms && SysTick->VAL / ticks_per_us >= us_until_ms) {}
 }
+
+#if CIRCUITPY_FULL_BUILD
+uint64_t common_hal_time_monotonic_ns(void) {
+    uint32_t ticks_per_us = common_hal_mcu_processor_get_frequency() / 1000 / 1000;
+
+    // We disable interrupts to prevent ticks_ms from changing while we grab it.
+    common_hal_mcu_disable_interrupts();
+    uint32_t tick_status = SysTick->CTRL;
+    uint32_t current_us = SysTick->VAL;
+    uint32_t tick_status2 = SysTick->CTRL;
```

Comment:
I think `tick_status` and `tick_status2` need to be `volatile` to ensure the compiler does not optimize away the second assignment.

Reply:
also fix in `current_tick()` above. Give there's a lot of copy-pasta code here, could the core part of the repeated code be refactored into a single routine? Also, it works for all ARM chips, so it could be shared among ports, except for Spresense.

Reply:
Agree about the desire to share code here.  I started to factor out the tick code in #2297 but didn't get to sharing the code in `current_tick`.

The reason the code from `current_tick` is not shared with the new `common_hal_time_monotonic_ns` is that the systick timer is finer grained than 1us (I think it's 15.625ns on nRF at 64MHz for instance), but `current_tick` reports in us.  The extra precision is probably not *actually useful*, so if it's preferable I can write the one in terms of the other.

I am unsure about the requirement to volatile-qualify tick_status / tick_status2.  Can you tell me more about the rules that are followed in CP?  My existing knowledge leads me to expect that as the register itself (SysTick->CTRL) is volatile qualified, and that on its own is enough to prevent the compiler from reordering the loads or eliminating repeated loads.

Reply:
Ah yes, `SysTick_Type` `volatile`, through a set of macros. I was doing a simple grep that didn't make that obvious.